### ISSUE TITLE
Opening PolyLineM Shapefile fails to return file type unsupported error

### ISSUE THREAD

**Issue by hasecbinusr** [pysal/pysal#932]
- Platform information:  Win 7 64bit (nt win32)
- Python version: 3.6.1 | packaged by conda-forge | (default, Mar 24 2017, 01:02:36) [MSC v.1900 64 bit (AMD64)]
- SciPy version: 0.19.0
- NumPy version: 1.11.3

Attempted to open an ESRI Shapefile with M attributes (route measure):
```python
roads = ps.open('roads.shp')
```
Resulted in the following error:
```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-10-4d9924bd5377> in <module>()
----> 1 roads = ps.open('roads.shp')

...\lib\site-packages\pysal\core\IOHandlers\pyShpIO.py in __init__(self, *args, **kwargs)
     66         self.dataObj = None
     67         if self.mode == 'r' or self.mode == 'rb':
---> 68             self.__open()
     69         elif self.mode == 'w' or self.mode == 'wb':
     70             self.__create()

...\lib\site-packages\pysal\core\IOHandlers\pyShpIO.py in __open(self)
     81         self.bbox = self.dataObj.bbox
     82         try:
---> 83             self.type = STRING_TO_TYPE[self.dataObj.type()]
     84         except KeyError:
     85             raise TypeError('%s does not support shapes of type: %s'

...\lib\site-packages\pysal\core\util\shapefile.py in type(self)
    320 
    321     def type(self):
--> 322         return self.shape.String_Type
    323 
    324     def __next__(self):

AttributeError: type object 'PolyLineM' has no attribute 'String_Type'
```
PySAL can open the shapefile after stripping the M attributes.

Interestingly, the script should call to line 754 and return the user with an informative error that this filetype is not supported yet:
```python
754  class PolyLineM:
755      def __init__(self):
756          raise NotImplementedError("No PolyLineM Support at this time.")
```
Perhaps attempting to load any of the unsupported file types will result in the same error at line 322 as above?

**Comment by ljwolf** [comment]
Thanks, that looks like a straghtfoward fix. Since we're not actually instantiating the object, the `NotImplementedError` never raises.

**Comment by hasecbinusr** [comment]
@ljwolf, what's going on here that's causing the error?

**Comment by ljwolf** [comment]
The shapefile reader is attempting to initialize via `__open`. 

It's looking up the types of the supported formats in a lookup table in line 17 using the type information stored in `self.dataObj.type()`. 

`self.dataObj` is a class that contains details needed to parse the many different possible shapefile types. One of those types that we don't support, `PolyLineM`, is definitely not in the lookup table in l17. But, to even **determine** the type of `self.dataObj`, it runs `self.dataObj.type()`, since the lookup table is keyed on that. 

`type()` looks up the attribute `String_Type` on the class you've identified in line 754. Since this class is never initialized, it never has a `String_Type` attribute. Since it doesn't, this raises an `AttributeError`.

If this `AttributeError` weren't raised (like, if `PolyLineM` had a `String_Type` attribute or raised its own error when accessed), then the **intended** error, a `KeyError` would activate from failing to find `PolyLineM` in `STRING_TO_TYPE` on line 17, and the exception caught correctly. In addition, if we only included shape types we supported in the lookup table `SHAPE_TYPES` of line 195 of `pysal/core/util/shapefile.py` or `TYPE_DISPATCH` on line 773 of the same file, the correct exception would be caught.

Instead, the `dataObj.__init__` completes fine & links to a class `PolyLineM` that never gets initialized. Then, `PolyLineM.shape.String_Type` is accessed, and, since it was never initialized, it cannot raise the `NotImplementedError` in its `__init__` method.

**Comment by hasecbinusr** [comment]
Thanks for the detailed explanation!  I'm new to this, so please excuse my ignorance... Would a fix for this be to include `String_Type = "ARCM"` in `class PolyLineM` or add in a try/except in `def type(self)`?

**Comment by ljwolf** [comment]
I think we would need to add `String_Type = "ARCM"` or `PolyLineM` (not sure which is appropriate) to the subclass, so that the `except` clause, which also looks at `dataObj.type()` executes correctly.

**Comment by hasecbinusr** [comment]
Resolved in `dev` branch with PR #934
